src/event-log/event-log-level.ts

```diff
@@ -0,0 +1,80 @@
+import type { ULID } from 'ulid';
+import type { Event, EventLog } from './event-log.js';
+
+import { base32crockford } from '@scure/base';
+import { Encoder } from '../utils/encoder.js';
+import { monotonicFactory } from 'ulid';
+import { createLevelDatabase, LevelWrapper } from '../store/level-wrapper.js';
+
+
+type EventLogLevelConfig = {
+ /**
+   * must be a directory path (relative or absolute) where
+   *  LevelDB will store its files, or in browsers, the name of the
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/IDBDatabase IDBDatabase} to be opened.
+  */
+  location: string,
+  createLevelDatabase?: typeof createLevelDatabase,
+};
+
+export class EventLogLevel implements EventLog {
+  config: EventLogLevelConfig;
+  level: LevelWrapper<string>;
+  ulid: ULID;
+
+  constructor(config: EventLogLevelConfig = { location: 'EVENTLOG' }) {
+    this.config = {
+      createLevelDatabase,
+      ...config,
+    };
+
+    this.level = new LevelWrapper<string>({ ...this.config, valueEncoding: 'utf8' });
+    this.ulid = monotonicFactory();
+  }
+
+  async open(): Promise<void> {
+    // `db.open()` is automatically called by the database constructor.  We're calling it explicitly
+    // in order to explicitly catch an error that would otherwise not surface until another method
+    // like `db.get()` is called.  Once `db.open()` has then been called, any read & write
+    // operations will again be queued internally until opening has finished.
+    return this.level.open();
+  }
+
+  close(): Promise<void> {
+    return this.level.close();
+  }
+
+  clear(): Promise<void> {
+    return this.level.clear();
+  }
+
+  async append(tenant: string, messageCid: string): Promise<string> {
+    const hashedTenant = this.hashTenant(tenant);
+    const tenantEventLog = await this.level.partition(hashedTenant);
+
+    const watermark = this.ulid();
+    await tenantEventLog.put(watermark, messageCid);
+
+    return watermark;
```

Comment:
In the context of event streams, a watermark is a technique used to track the progress of events within a stream and to manage event processing.

In event processing, events are typically processed in batches rather than one at a time. A watermark is a timestamp that represents the last event in a batch of events that have been processed. As new events are added to the stream, the watermark is updated to represent the most recent processed event.

The purpose of a watermark is to provide a way to manage the progress of event processing, to prevent event duplication, and to provide a guarantee of event ordering. By using a watermark, event processing systems can ensure that they have processed all events up to a certain point in time before moving on to process newer events.

down to change the term/word to something else if you've got something in mind!